kernel/thread/thread.c

```diff
@@ -613,6 +627,20 @@ int thd_set_prio(kthread_t *thd, prio_t prio) {
 /*****************************************************************************/
 /* Scheduling routines */
 
+static void thd_update_cpu_time(kthread_t *thd) {
+    if(perf_cntr_timer_enabled()) {
+
+        const uint64_t ns = perf_cntr_timer_ns();
+
+        if(thd_current) {
+            thd_current->cpu_time.total +=
+                ns - thd_current->cpu_time.scheduled;
```

Comment:
Should there be some check for the scheduled being 0/uninitialized? It would seem that since perf_cntr_timer_enabled is being treated as run-time variable and not tied to some mode that you'd get really weird garbage for all threads if you enabled it on-the-fly. 

Or maybe it's not meant to really be runtime settable and it's meant more to be something that is static during the course of execution?

Reply:
It was originally intended to be static throughout the course of execution, like how you can't actually modify the CPU clock as you can the REALTIME clock on Linux... so I was operating under the assumption that if you had chosen to reconfigure or claim that first perf counter that KOS uses for this timing stuff, that it was understood that timing would be off and you would no longer be able to measure per-thread or per-process CPU time...

With this in mind, do you think I'm better off hard zeroing it out here if they did mess with that performance counter? Or leave it as its last valid value, which I think is what this should do?

Reply:
I think the thing that I missed in my thinking on it was that perf counters are by default enabled and that happens directly before threads are initted. So I was a bit concerned about something like the default thread having an uninitialized value for cpu_time.scheduled/total. But since new threads are 0-initted that would only be an issue when someone has disabled perf counters, created new threads, then reenabled perf counters. I think that'd be on the user, and am pretty sure this would be safe even if `(ns-scheduled) <0` it'd just give garbage that they can deal with.

What's the reason though for not just allowing the perfcounters to use the fallback value out of timer_ns_gettime64 if they happen to be disabled? I'm guessing it has something to do with the differences between those systems (counting nops?) but it would seem like that it might be preferable to get slightly incorrect info from them than the wildly incorrect from stopping the counting? If that could work as the backup then you could do away with all the checks for the perf counters being enabled.

Reply:
Yeah, actually, everything you've said is correct here.

The reason I wasn't just falling back to `timer_ns_gettime64()` was simply because of fact the perf counters in this mode were not counting `SLEEP` instructions... but... I guess this may still be preferable to just garbage data... good point.